### PR TITLE
Sagemaker PyTorch Distributed Model Parallel GPT2 Updates

### PR DESCRIPTION
**Pull request by SuhitK**
*Issue #, if available:*

*Description of changes:*

- Add functionality to optionally save full optimizer state
- Added some memory tracking features
- Switched to using official HuggingFace estimator with PyTorch version 1.10

*Testing done:*
Tested end-to-end for `gpt-xl` model on 1 p4d.24xlarge instance. Training job was successfully completed.
## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [ x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-sagemaker-examples/blob/master/CONTRIBUTING.md) doc and adhered to the example notebook best practices
- [ x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/amazon-sagemaker-examples/blob/master/README.md)
- [x ] I have tested my notebook(s) and ensured it runs end-to-end
- [ ] I have linted my notebook(s) and code using `tox -e black-format,black-nb-format`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py [hunk 1]
```diff
@@ -1081,26 +1132,34 @@ def main():
...
-            model,
-            optimizer,
-            static_loss_scale=None,
-            dynamic_loss_scale=True,
-            use_smp=True,
-            dynamic_loss_args={"scale_window": 1000, "min_scale": 1, "delayed_shift": 2},
-            params_have_main_grad=args.fp32_grad_accumulation > 0,
-            shard_optimizer_state=args.shard_optimizer_state > 0,
-        )
+            optimizer = FP16_Optimizer(
```
Comment (rahul003) [inline training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py:1141]: Indentation needs to be fixed

File: training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py [hunk 2]
```diff
@@ -1047,7 +1098,7 @@ def main():
 
     if args.manual_partition:
         print(f"Manual partition enabled")
-        # evenly distribute layers across all partitions
+            # evenly distribute layers across all partitions
```
Comment (rahul003) [inline training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py:1101]: indentation

Comment (rahul003) [inline training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py:1101]: indentation

File: training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py [hunk 3]
```diff
@@ -588,15 +605,19 @@ def should_record():
                 loss_metric = loss.item()
 
             if args.fp16:
-                optimizer.update_master_grads()
-                optimizer.clip_master_grads(args.grad_clip)
-                optimizer.step()
-                overflow = optimizer.overflow
+                    optimizer.update_master_grads()
```
Comment (rahul003) [inline training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py:608]: indentation

Comment (rahul003) [inline training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py:608]: indentation

File: training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py [hunk 4]
```diff
@@ -97,7 +97,7 @@ def train_step(model, optimizer, input_ids, attention_mask, args):
     else:
         loss = model(input_ids=input_ids, attention_mask=attention_mask, labels=input_ids)["loss"]
     if args.fp16:
-        optimizer.backward(loss, update_master_grads=False)
+            optimizer.backward(loss, update_master_grads=False)
```
Comment (rahul003) [inline training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py:100]: There's more cases that I haven't pointed to as well, please review the file's indentation changes

Comment (rahul003) [inline training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py:100]: There's more cases that I haven't pointed to as well, please review the file's indentation changes

**Review by rahul003 — CHANGES_REQUESTED** [review]

File: training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb [hunk 5]
```diff
@@ -561,9 +561,9 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "### Create a SageMaker PyTorch Estimator\n",
+    "### Create a Sagemaker HuggingFace Estimator\n",
```
Comment (mchoi8739) [inline training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:564]:
```suggestion
    "### Create a SageMaker HuggingFace Estimator\n",
```

File: training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py [hunk 6]
```diff
@@ -223,7 +223,9 @@ def save(
     if lr_scheduler is not None:
         save_dict["lr_scheduler"] = lr_scheduler.state_dict()
     if partial:
-        save_dict["model"] = model.local_state_dict()
+        if args.gather_if_shard > 0 or smp.rdp_rank() == 0:
```
Comment (mchoi8739) [inline training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py:226]:
```suggestion
        # SMP modification required if you use optimizer state sharding
        if args.gather_if_shard > 0 or smp.rdp_rank() == 0:
```

File: training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py [hunk 7]
```diff
@@ -253,8 +255,8 @@ def save(
             else:
                 print("Skipping saving of full optimizer state")
 
-    if (smp.rdp_rank() == 0 and partial) or smp.rank() == 0:
-        smp.save(save_dict, output_save_file, partial=partial)
+    if not args.gather_if_shard or (smp.rdp_rank() == 0 and partial) or smp.rank() == 0:
```
Comment (mchoi8739) [inline training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py:258]:
```suggestion
    # SMP modification required: criteria for checkpointing the zeroth rank for pipeline parallelism, checkpointing the zeroth reduced data parallel rank for tensor parallelism, and preventing checkpointing if optimizer state sharding is enabled
    if not args.gather_if_shard or (smp.rdp_rank() == 0 and partial) or smp.rank() == 0:
```

File: training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb [hunk 8]
```diff
@@ -76,7 +76,7 @@
     "\n",
```
Comment (mchoi8739) [inline training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:76]: <p>I recommend you to mark which hyperparams are added for SMP config. Maybe L16-L19? </p><p>Also add reference to https://sagemaker.readthedocs.io/en/stable/api/training/smd_model_parallel_general.html</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3311/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (mchoi8739) [inline training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:76]: <p>Do we really need to have this section? We don't use any feature of Experiment in the notebook, only creating an Experiment trial that's basically unused.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3311/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='20'/>

Comment (mchoi8739) [inline training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:76]: <p>Again, not sure if we really need to have the Experiments trial, even for CI. If not needed, this makes the notebook to have an unnecessary step and overly complicated. </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3311/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='34'/>

Comment (mchoi8739) [inline training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:76]: <p>In the itemized list, we stated the following:</p><p> </p><blockquote><code>train_gpt_simple.py</code>: This is an entrypoint script that is passed to the Pytorch estimator in the notebook instructions. This script is responsible for end to end training of the GPT-2 model with SMP. <strong>The script has additional comments at places where the SMP API is used.</strong></blockquote><p> </p><p>Therefore, in the training script, please make sure adding commentary lines to everywhere you have SMP modifications.</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3311/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (SuhitK) [reply training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:76]: Agree, will remove it

Comment (SuhitK) [reply training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:76]: Comments were already added but were marked with smdistributed tag, I've updated the comments to all have the same tag of 'SMP modification'.

Comment (SuhitK) [reply training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:76]: Updated

Comment (mchoi8739) [reply training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:76]: <p>I think you can remove everything but the last line.</p><p> </p><pre class="ql-syntax" spellcheck="false">smp_estimator.fit(inputs=data_channels, logs=True)
</pre>

Comment (SuhitK) [reply training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:76]: <p>Updated</p>

**Review by mchoi8739 — APPROVED** [review]

File: training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb [hunk 9]
```diff
@@ -74,12 +74,10 @@
     "%%time\n",
```
Comment (mchoi8739) [inline training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:74]: <p>For <code>train_gpt_simple.py</code>,change:</p><p>the Pytorch estimator -&gt; the HuggingFace estimator </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3311/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

**Review by mchoi8739 — DISMISSED** [review]

File: training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb [hunk 10]
```diff
@@ -17,7 +17,7 @@
     "\n",
```
Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:17]: <p>Branding nits: Huggingface -&gt; Hugging Face	(find all occurrences), <code>Pytorch</code> -&gt; <code>PyTorch</code>   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3311/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (jkroll-aws) [inline training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:17]: <p>Branding: <code>Sagemaker</code> -&gt; <code>SageMaker</code>   </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/aws/amazon-sagemaker-examples/pull/3311/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (SuhitK) [reply training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:17]: Updated

Comment (SuhitK) [reply training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:17]: Updated

File: training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py [hunk 11]
```diff
@@ -253,8 +256,9 @@ def save(
             else:
                 print("Skipping saving of full optimizer state")
 
-    if (smp.rdp_rank() == 0 and partial) or smp.rank() == 0:
-        smp.save(save_dict, output_save_file, partial=partial)
+    # SMP modification: criteria for checkpointing the zeroth rank for pipeline parallelism, checkpointing the zeroth reduced data parallel rank for tensor parallelism, and preventing checkpointing if optimizer state sharding is enabled
```
Comment (rahul003) [inline training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py:259]: line is too long, hard to read.

File: training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py [hunk 12]
```diff
@@ -634,6 +657,8 @@ def should_record():
                 out_path = os.path.join(args.checkpoint_dir, base_path)
                 total_ckpts = total_steps // args.checkpoint_freq
 
+                delete_oldest_ckpt(args, delete_on_rank0_only=args.use_fsx > 0)
```
Comment (rahul003) [inline training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py:660]: This isn't already in the file elsewhere? We didn't have this support?

File: training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb [hunk 13]
```diff
@@ -307,18 +305,27 @@
...
-    "    \"shard_optimizer_state\": 1,\n",
-    "    \"activation_checkpointing\": 1,\n",
-    "    \"activation_strategy\": \"each\",\n",
-    "    \"optimize\": \"speed\",\n",
     "    \"use_wiki_data\": 1,\n",
     "    # below flag loads model and optimizer state from checkpoint_s3_uri\n",
     "    # 'load_partial': 1,\n",
     "}\n",
     "\n",
+    "# Add parameters required by SMP config.\n",
```
Comment (rahul003) [inline training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:313]: Why are these separated like this from above set?

Comment (SuhitK) [reply training/distributed_training/pytorch/model_parallel/gpt2/smp-train-gpt-simple.ipynb:313]: This was done based on the suggestion to mark which parameters correspond to SMP config: https://github.com/aws/amazon-sagemaker-examples/pull/3311#discussion_r878472254

File: training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py (hunk 12 above)
Comment (SuhitK) [reply training/distributed_training/pytorch/model_parallel/gpt2/train_gpt_simple.py:660]: It was, just has been moved from L663 to here.

**Review by jkroll-aws — APPROVED** [review]

**Pull request merged**